Shard builder <> proposer separation (#2486)
ethereum/eth2.0-specs: specs/sharding/p2p-interface.md

```

The following validations MUST pass before forwarding the `signed_blob`,
on the horizontal subnet or creating samples for it. Alias `blob = signed_blob.message`.

- _[IGNORE]_ The `blob` is not from a future slot (with a `MAXIMUM_GOSSIP_CLOCK_DISPARITY` allowance) --
  i.e. validate that `blob.slot <= current_slot`
  (a client MAY queue future blobs for processing at the appropriate slot).
- _[IGNORE]_ The `blob` is new enough to still be processed --
  i.e. validate that `compute_epoch_at_slot(blob.slot) >= get_previous_epoch(state)`
- _[REJECT]_ The shard blob is for an active shard --
  i.e. `blob.shard < get_active_shard_count(state, compute_epoch_at_slot(blob.slot))`
- _[REJECT]_ The `blob.shard` MUST have a committee at the `blob.slot` --
  i.e. validate that `compute_committee_index_from_shard(state, blob.slot, blob.shard)` doesn't raise an error
- _[REJECT]_ The shard blob is for the correct subnet --
  i.e. `compute_subnet_for_shard_blob(state, blob.slot, blob.shard) == subnet_id`
- _[IGNORE]_ The blob is the first blob with valid signature received for the `(blob.proposer_index, blob.slot, blob.shard)` combination.

> - _[REJECT]_ The blob is not too large, the data MUST NOT be larger than the SSZ list-limit, and a client MAY be more strict.

Reviewer: ```suggestion
- _[REJECT]_ The blob is not too large -- the data MUST NOT be larger than the SSZ list-limit, and a client MAY apply stricter bounds.
```
```suggestion
- _[REJECT]_ The blob is not too large, the `blob.body.data` MUST NOT be larger than the SSZ list-limit, and a client MAY be more strict.
```

- _[REJECT]_ The `blob.body.data` MUST NOT contain any point `p >= MODULUS`. Although it is a `uint256`, not the full 256 bit range is valid.

> - _[REJECT]_ The blob builder exists and has sufficient balance to back the fee payment.

Reviewer: ```suggestion
- _[REJECT]_ The blob builder defined by `blob.builder_index` exists and has sufficient balance to back the fee payment.
```

> - _[REJECT]_ The blob signature is valid for the aggregate of proposer and builder, `signed_blob.signature`,

Reviewer: ```suggestion
- _[REJECT]_ The blob signature, `signed_blob.signature`, is valid for the aggregate of proposer and builder --
```

  i.e. `bls.FastAggregateVerify([builder_pubkey, proposer_pubkey], blob_signing_root, signed_blob.signature)`.
- _[REJECT]_ The blob is proposed by the expected `proposer_index` for the blob's `slot` and `shard`,
  in the context of the current shuffling (defined by `blob.body.beacon_block_root`/`slot`).
  If the `proposer_index` cannot immediately be verified against the expected shuffling,
  the blob MAY be queued for later processing while proposers for the blob's branch are calculated --
  in such a case _do not_ `REJECT`, instead `IGNORE` this message.

#### Global topics

There are three additional global topics for Sharding.

> - `shard_blob_header`: co-signed headers, to be included on-chain, and signaling builders to publish full data.

Reviewer: ```suggestion
- `shard_blob_header`: co-signed headers to be included on-chain and to serve as a signal to the builder to publish full data.
```

- `shard_blob_tx`: builder-signed headers, also known as "data transaction".

> - `shard_proposer_slashing`: slashings of duplicate shard proposals

Reviewer: ```suggestion
- `shard_proposer_slashing`: slashings of duplicate shard proposals.
```

##### `shard_blob_header`

Shard header data, in the form of a `SignedShardBlobHeader` is published to the global `shard_blob_header` subnet.

> Shard blob headers select shard blob bids by builders,

Reviewer: ```suggestion
Shard blob headers select shard blob bids by builders
```

and should be timely to ensure builders can publish the full shard blob before subsequent attestations.

> The following validations MUST pass before forwarding the `signed_blob_header` on the network. Alias `header = signed_blob_header.message`

Reviewer: ```suggestion
The following validations MUST pass before forwarding the `signed_blob_header` on the network. Alias `header = signed_blob_header.message`.
```

- _[IGNORE]_ The `header` is not from a future slot (with a `MAXIMUM_GOSSIP_CLOCK_DISPARITY` allowance) --
  i.e. validate that `header.slot <= current_slot`
  (a client MAY queue future headers for processing at the appropriate slot).
- _[IGNORE]_ The header is new enough to still be processed --
  i.e. validate that `compute_epoch_at_slot(header.slot) >= get_previous_epoch(state)`
- _[REJECT]_ The shard header is for an active shard --
  i.e. `header.shard < get_active_shard_count(state, compute_epoch_at_slot(header.slot))`
- _[REJECT]_ The `header.shard` MUST have a committee at the `header.slot` --
  i.e. validate that `compute_committee_index_from_shard(state, header.slot, header.shard)` doesn't raise an error
- _[IGNORE]_ The header is the first header with valid signature received for the `(header.proposer_index, header.slot, header.shard)` combination.

> - _[REJECT]_ The blob builder exists and has sufficient balance to back the fee payment.

Reviewer: ```suggestion
- _[REJECT]_ The blob builder defined by `blob.builder_index` exists and has sufficient balance to back the fee payment.
```

> - _[REJECT]_ The header signature is valid for the aggregate of proposer and builder, `signed_blob_header.signature`,

Reviewer: ```suggestion
- _[REJECT]_ The header signature, `signed_blob_header.signature`, is valid for the aggregate of proposer and builder --
```

  i.e. `bls.FastAggregateVerify([builder_pubkey, proposer_pubkey], blob_signing_root, signed_blob_header.signature)`.
- _[REJECT]_ The header is proposed by the expected `proposer_index` for the blob's `header.slot` and `header.shard`
  in the context of the current shuffling (defined by `header.body_summary.beacon_block_root`/`slot`).
  If the `proposer_index` cannot immediately be verified against the expected shuffling,
  the blob MAY be queued for later processing while proposers for the blob's branch are calculated --
  in such a case _do not_ `REJECT`, instead `IGNORE` this message.

##### `shard_blob_tx`

> Shard data-transactions, in the form of a `SignedShardBlobHeader` is published to the global `shard_blob_tx` subnet.

Reviewer: ```suggestion
Shard data-transactions in the form of a `SignedShardBlobHeader` are published to the global `shard_blob_tx` subnet.
```

> Shard data-transactions, in the form of a `SignedShardBlobHeader` is published to the global `shard_blob_tx` subnet.

Reviewer 2: Maybe I'm missing something, shouldn't this be `SignedShardBlob` instead of `SignedShardBlobHeader`.  In order to publish the full data, don't we need `data` instead of `data_root` ?

Reviewer 3: @terencechain full data publishing is now the builder's duty. As soon as he sees his `SignedShardBlobHeader` on `blob_headers` topic signed by proposer (effectively an aggregate signature: builder + proposer) he has strong incentive to publish the corresponding `SignedShardBlob` as soon as possible.
@protolambda also mentioned that hiding actual blob data until proposer acceptance may potentially add more security as proposer would not be able to filter/withhold blobs based on their content

These shard blob headers are signed solely by the blob-builder.

> The following validations MUST pass before forwarding the `signed_blob_header` on the network. Alias `header = signed_blob_header.message`

Reviewer: ```suggestion
The following validations MUST pass before forwarding the `signed_blob_header` on the network. Alias `header = signed_blob_header.message`.
```

> - _[IGNORE]_ The `header` is not from a future slot (with a `MAXIMUM_GOSSIP_CLOCK_DISPARITY` allowance) --

Reviewer: ```suggestion
- _[IGNORE]_ The header is not from a future slot (with a `MAXIMUM_GOSSIP_CLOCK_DISPARITY` allowance) --
```

>   i.e. validate that `header.slot <= current_slot`

Reviewer: The writing of this condition seems a bit too strong. Normally, I'm going to broadcast for inclusion in slot `N` in slot `N-1` (or earlier) so *almost always* it will be for a future slot

Reviewer: In fact, I"m unsure if there should be an upper bound in the rules here. Maybe a sanity bound? But then, it's more on the data-tx mempool management to flush things that are stale, invaluable, or no longer able to pay fees

>   (a client MAY queue future headers for processing at the appropriate slot).

Reviewer: And you precisely *want* headers/data-txs from the future to put in your mempool. but also want to forward them as well.

- _[IGNORE]_ The header is new enough to still be processed --

>   i.e. validate that `compute_epoch_at_slot(header.slot) >= get_previous_epoch(state)`

Reviewer: Whereas, this condition seems a bit too lax! things even a couple of slots ago are no longer valuable at all! unless I'm doing some sort of re-org attack and re-packing the chain..

- _[REJECT]_ The shard header is for an active shard --
  i.e. `header.shard < get_active_shard_count(state, compute_epoch_at_slot(header.slot))`

> - _[REJECT]_ The `header.shard` MUST have a committee at the `header.slot` --

Reviewer: We repeat these conditions 3 times. Essentially the "shard blob header" validations. Should we consider wrapping them to reduce copy pasta issues?

  i.e. validate that `compute_committee_index_from_shard(state, header.slot, header.shard)` doesn't raise an error

> - _[IGNORE]_ The header is the first header with valid signature received for the `(header.builder_index, header.slot, header.shard)` combination.

Reviewer: this should actually probably be the *highest fee payment* for that combo. So if less than or equal, drop. but a builder should be able to re-send the same or something different for shard/slot that has a higher fee.

> - _[REJECT]_ The blob builder exists and has sufficient balance to back the fee payment.

Reviewer: should this consider data-txs in the mempool for the same builder and earlier slots?
that is, assume they *will* make it in and this maybe won't be able be to be paid for? or too strict

Reviewer: ```suggestion
- _[REJECT]_ The blob builder, define by `header.builder_index`, exists and has sufficient balance to back the fee payment.
```

Author: > should this consider data-txs in the mempool for the same builder and earlier slots?
that is, assume they will make it in and this maybe won't be able be to be paid for? or too strict

We should probably address this as part of the fix for parallel fee payments. Doesn't matter if it's in parallel or in-flight on an earlier slot, header fees need to be assured.

> - _[IGNORE]_ The header fee SHOULD be higher than previously seen headers for `(header.slot, header.shard)`, from any builder.

Reviewer: ah, interesting... drop data-txs of lower fees for that shard/slot combo because they are so *pointedly* valuable.

I see that. See previous comment about builder being able to supersede their previous data-tx with higher fee. maybe combine with this statement.

Author: We would enable a builder to spam the network if we allow a new data-tx for the same `(slot, shard)` combination, with a higher priority fee. I.e. it's pointless if a builder spams 1.001, 1.002, ...., 1.999 for priority fee. We could define a threshold for replacement to counter spam, but right now I prefer to keep it simple.

>   Propagating nodes MAY increase fee increments in case of spam.

Reviewer: I'm not sure I understand this. As in, must be `> prev_max_fee + DELTA` rather than just `> prev_max_fee`?

Author: >  As in, must be > prev_max_fee + DELTA rather than just > prev_max_fee?

Yes (although it should probably account for base fee and priority fee increments separately)

Similarly to a single builder spamming replacement transactions that only slightly increase the fee, multiple builders could coordinate to spam. E.g. 100 builders publish a transaction, all with slightly increasing fee. As a network participant, it may not be worth propagating every tiny fee increment, as long as one of the highest few get through.

> - _[REJECT]_ The header signature is valid for ONLY the builder, `signed_blob_header.signature`,

Reviewer: ```suggestion
- _[REJECT]_ The header signature, `signed_blob_header.signature`, is valid for ONLY the builder --
```

- _[IGNORE]_ The header is the first header with valid signature received for the `(header.builder_index, header.slot, header.shard)` combination.
- _[REJECT]_ The blob builder exists and has sufficient balance to back the fee payment.
- _[IGNORE]_ The header fee SHOULD be higher than previously seen headers for `(header.slot, header.shard)`, from any builder.
  Propagating nodes MAY increase fee increments in case of spam.
- _[REJECT]_ The header signature is valid for ONLY the builder, `signed_blob_header.signature`,
  i.e. `bls.Verify(builder_pubkey, blob_signing_root, signed_blob_header.signature)`. The signature is not an aggregate with the proposer.
- _[REJECT]_ The header is designated for proposal by the expected `proposer_index` for the blob's `header.slot` and `header.shard`
  in the context of the current shuffling (defined by `header.body_summary.beacon_block_root`/`slot`).
  If the `proposer_index` cannot immediately be verified against the expected shuffling,
  the blob MAY be queued for later processing while proposers for the blob's branch are calculated --
  in such a case _do not_ `REJECT`, instead `IGNORE` this message.

> 

Reviewer 3: BTW do we actually need `proposer_index` field at all (in `ShardBlob` and `ShardBlobHeader`)? It is derivable from `(beacon_block_root, slot, shard)`
It seems no much sense for a _builder_ to calculate the designated proposer index when creating a shard tx - he shouldn't actually care who would propose his blob.

Author: The builder already commits the transaction to a specific shard/slot. And if the builder does not win, the blob is very likely stale, and shouldn't be applied anyways later on. Computing the proposer index as builder should not be relatively much overhead.

Meanwhile, just like in beacon-blocks, having the proposer index in the signed message is useful for various things (easy slashing mechanism, optimized p2p validation order, etc.). It may change if we get single-secret-leader election, and builders don't know the shard proposers, but until then it seems more optimal to include it.

The following validations MUST pass before forwarding the `signed_blob`,
on the horizontal subnet or creating samples for it. Alias `blob = signed_blob.message`.

> - _[IGNORE]_ The `blob` is published 1 slot early or later (with a `MAXIMUM_GOSSIP_CLOCK_DISPARITY` allowance) --

Reviewer: One slot because the proposer moves at N - 1/3, meaning the builder should move *immediately* which might be marginally before N?

Author: exactly
